### PR TITLE
Caching for Quickcost and bugfix for CalculateProducts

### PR DESCRIPTION
**Pull request by francoisfeugeas**
Added caching for Chronopost's QuickCost webservice and fixed CalculateProducts behaviour when cart weight == 0

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
Hi @francoisfeugeas,

I'm sorry, but this PR is invalid. You have conflicts marks in your suggested code.
You have to remove them before we can validate this PR.

Best regards

**Comment by francoisfeugeas** [comment]
Sorry about that @Quetzacoalt91 - should be good to go now

**Comment by Quetzacoalt91** [comment]
Thanks, I will review it ASAP

**Comment by Quetzacoalt91** [comment]
Hi @francoisfeugeas,

I see that you create a table during the install of this new version.
You also have to manage the creation during the upgrade, and for this, you need to create an upgrade file.

If needed, see the documentation to get more information: http://doc.prestashop.com/display/PS15/Auto-updating+modules

**Comment by francoisfeugeas** [comment]
Hi @Quetzacoalt91 ,

The upgrade file is created now.

File: chronopost.php [hunk 1]
```diff
@@ -875,9 +897,29 @@ public function getOrderShippingCost($cart, $shipping_cost)
...
 
+		if ($res->return->amountTTC != 0)
+		{
+			if(empty($cache))
+			{
+				DB::getInstance()->query('INSERT INTO '._DB_PREFIX_.'chrono_quickcost_cache (product_code, arrcode, weight, price, last_updated) VALUES (
+						"'.$productCode.'",
+						"'.$arrcode.'",
+						"'.$cart->getTotalWeight().'",
+						"'.$res->return->amountTTC.'",
```
Comment (Quetzacoalt91) [inline chronopost.php]: Can you protect a little bit more the data sent to your database ? with an `(int)` cast for numbers and `pSql(...)` for strings

File: chronopost.php [hunk 2]
```diff
@@ -875,9 +897,29 @@ public function getOrderShippingCost($cart, $shipping_cost)
...
+						"'.time().'")
+				');
+			}
+			else 
+				DB::getInstance()->query('UPDATE '._DB_PREFIX_.'chrono_quickcost_cache SET price="'.$res->return->amount.'", last_updated="'.time().'
+					WHERE arrcode = "'.pSQL($arrcode).'" && product_code="'.$productCode.'" && weight="'.$cart->getTotalWeight().'"
```
Comment (Quetzacoalt91) [inline chronopost.php]: And also here

File: config.xml [hunk 3]
```diff
@@ -1,13 +0,0 @@
...
-	<displayName><![CDATA[Chronopost and ChronoRelais]]></displayName>
-	<version><![CDATA[3.6.3]]></version>
-	<description><![CDATA[Manage Chronopost and Chronopost Pickup relay]]></description>
-	<author><![CDATA[Oxileo for Chronopost]]></author>
-	<tab><![CDATA[shipping_logistics]]></tab>
-	<confirmUninstall><![CDATA[Remember, once this module is uninstalled , you won't be able to edit Chronopost waybills or propose Pickup delivery point to your customers. Are you sure you wish to proceed?]]></confirmUninstall>
-	<is_configurable>1</is_configurable>
-	<need_instance>1</need_instance>
-	<limited_countries></limited_countries>
-</module>
```
Comment (Quetzacoalt91) [inline config.xml]: We use the file config.xml for the cache. Can you please bring it back ?

File: upgrade/Upgrade-3.4.6.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?php
+if (!defined('_PS_VERSION_'))
+    exit;
+
+function upgrade_module_3_4_6($object)
```
Comment (Quetzacoalt91) [inline upgrade/Upgrade-3.4.6.php]: I guess your function name is wrong, because I see that the new version number is 3.6.4 in your main class. ;)

**Comment by francoisfeugeas** [comment]
@Quetzacoalt91 fixed all of your remarks.

**Comment by Quetzacoalt91** [comment]
Hi @francoisfeugeas,

Thanks, your PR has just been merged and the changes should be available from tomorrow.

Best regards

**Pull request merged**